The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6781306826/job/18431382379) [(plain)](https://github.com/rust-lang/rust/commit/b060c18c053d50b63cf47b04f6c4ac674817126d/checks/18431382379/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
diff of stderr:

18   --> $DIR/chain-method-call-mutation-in-place.rs:3:72
19    |
20 LL |     vec![1, 2, 3].into_iter().collect::<Vec<i32>>().sort_by_key(|i| i).sort();
-    |                   -----------                       ------------------ ^^^^ method not found in `()`
-    |                   |                                 |
-    |                   |                                 method `sort` is available on `&mut [i32]`
-    |                   method `sort` is available on `Vec<i32>`
+    |     -----------------------------------------------                    ^^^^ method not found in `()`
+    |     |
+    |     method `sort` is available on `&mut [i32]`
+    |     method `sort` is available on `Vec<i32>`
25    |
26 note: method `sort_by_key` modifies its receiver in-place, it is not meant to be used in method chains.


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/suggestions/chain-method-call-mutation-in-place/chain-method-call-mutation-in-place.stderr
---
--- stderr -------------------------------
error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/suggestions/chain-method-call-mutation-in-place.rs:2:23
   |
LL |     let x: Vec<i32> = vec![1, 2, 3].into_iter().collect::<Vec<i32>>().sort_by_key(|i| i); //~ ERROR mismatched types
   |            --------   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Vec<i32>`, found `()`
Build completed unsuccessfully in 0:12:56
   |            expected due to this
   |
   = note: expected struct `Vec<i32>`
   = note: expected struct `Vec<i32>`
           found unit type `()`
note: method `sort_by_key` modifies its receiver in-place, it is not meant to be used in method chains.
   |
   |
LL |     let x: Vec<i32> = vec![1, 2, 3].into_iter().collect::<Vec<i32>>().sort_by_key(|i| i); //~ ERROR mismatched types
   |                                                                       ^^^^^^^^^^^ this call modifies its receiver in-place

error[E0599]: no method named `sort` found for unit type `()` in the current scope
   |
   |
LL |     vec![1, 2, 3].into_iter().collect::<Vec<i32>>().sort_by_key(|i| i).sort(); //~ ERROR no method named `sort` found for unit type `()` ...
   |     -----------------------------------------------                    ^^^^ method not found in `()`
   |     |
   |     method `sort` is available on `&mut [i32]`
   |     method `sort` is available on `Vec<i32>`
   |
note: method `sort_by_key` modifies its receiver in-place, it is not meant to be used in method chains.
   |
   |
LL |     vec![1, 2, 3].into_iter().collect::<Vec<i32>>().sort_by_key(|i| i).sort(); //~ ERROR no method named `sort` found for unit type `()` ...
   |                                                     ^^^^^^^^^^^ this call modifies its receiver in-place
error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/suggestions/chain-method-call-mutation-in-place.rs:7:5
   |
   |
LL | fn foo(mut s: String) -> String {
   |                          ------ expected `String` because of return type
LL |     s.push_str("asdf") //~ ERROR mismatched types
   |     ^^^^^^^^^^^^^^^^^^ expected `String`, found `()`
   |
note: method `push_str` modifies its receiver in-place
   |
   |
LL |     s.push_str("asdf") //~ ERROR mismatched types
   |     - ^^^^^^^^ this call modifies `s` in-place
   |     you probably want to use this value after calling the method...
   |     you probably want to use this value after calling the method...
   = note: ...instead of the `()` output of method `push_str`
error: aborting due to 3 previous errors

Some errors have detailed explanations: E0308, E0599.
For more information about an error, try `rustc --explain E0308`.
```

</details>

--- a/compiler/rustc_hir_typeck/src/method/suggest.rs
+++ b/compiler/rustc_hir_typeck/src/method/suggest.rs
@@ -481,7 +481,6 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 );
                 probe.is_ok()
             });
-
             self.note_internal_mutation_in_method(
                 &mut err,
                 rcvr_expr,
@@ -1240,7 +1239,40 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 }
             }
         }
+        // If an appropriate error source is not found, check method chain for possible candiates
+        if unsatisfied_predicates.is_empty() && let Mode::MethodCall = mode && let SelfSource::MethodCall(mut source_expr) = source {
+            let mut stack_methods = vec![];
+            while let hir::ExprKind::MethodCall(_path_segment, rcvr_expr, _args, method_span) =
+            source_expr.kind
+            {
+                 // Pop the matching receiver, to align on it's notional span
+                 if let Some(prev_match) = stack_methods.pop() {
+                    err.span_label(method_span, format!("{item_kind} `{item_name}` is available on `{prev_match}`"));
+                }
+                let rcvr_ty = self.resolve_vars_if_possible(
+                    self.typeck_results
+                        .borrow()
+                        .expr_ty_adjusted_opt(rcvr_expr)
+                        .unwrap_or(Ty::new_misc_error(self.tcx)),);
 
+                for _matched_method in self.probe_for_name_many(
+                    Mode::MethodCall,
+                    item_name,
+                    None,
+                    IsSuggestion(true),
+                    rcvr_ty,
+                    source_expr.hir_id,
+                    ProbeScope::TraitsInScope,) {
+                        // found a match, push to stack
+                        stack_methods.push(rcvr_ty);
+                }
+                source_expr = rcvr_expr;
+            }
+            // If there is a match at the start of the chain, add a label for it too!
+            if let Some(prev_match) = stack_methods.pop() {
+                err.span_label(source_expr.span, format!("{item_kind} `{item_name}` is available on `{prev_match}`"));
+            }
+        }
         self.note_derefed_ty_has_method(&mut err, source, rcvr_ty, item_name, expected);
         return Some(err);
     }

--- a/tests/ui/structs/method-chain-expression-failure.rs
+++ b/tests/ui/structs/method-chain-expression-failure.rs
@@ -1,0 +1,31 @@
+struct A;
+struct B;
+struct C;
+struct D;
+struct E;
+
+impl A {
+    fn b(&self) -> B { B }
+    fn foo(&self) {}
+}
+
+impl B {
+    fn c(&self) -> C { C }
+}
+
+impl C {
+    fn d(&self) -> D { D }
+    fn foo(&self) {}
+}
+
+impl D {
+    fn e(&self) -> E { E }
+}
+
+impl E {
+    fn f(&self) {}
+}
+fn main() {
+    A.b().c().d().e().foo();
+//~^ ERROR no method named `foo` found for struct `E` in the current scope
+}

--- a/tests/ui/structs/method-chain-expression-failure.stderr
+++ b/tests/ui/structs/method-chain-expression-failure.stderr
@@ -1,0 +1,15 @@
+error[E0599]: no method named `foo` found for struct `E` in the current scope
+  --> $DIR/method-chain-expression-failure.rs:29:23
+   |
+LL | struct E;
+   | -------- method `foo` not found for this struct
+...
+LL |     A.b().c().d().e().foo();
+   |     -     ---         ^^^ method not found in `E`
+   |     |     |
+   |     |     method `foo` is available on `&C`
+   |     method `foo` is available on `&A`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0599`.

--- a/tests/ui/suggestions/chain-method-call-mutation-in-place.stderr
+++ b/tests/ui/suggestions/chain-method-call-mutation-in-place.stderr
@@ -18,7 +18,10 @@ error[E0599]: no method named `sort` found for unit type `()` in the current sco
   --> $DIR/chain-method-call-mutation-in-place.rs:3:72
    |
 LL |     vec![1, 2, 3].into_iter().collect::<Vec<i32>>().sort_by_key(|i| i).sort();
-   |                                                                        ^^^^ method not found in `()`
+   |     -------------             ---------------------                    ^^^^ method not found in `()`
+   |     |                         |
+   |     |                         method `sort` is available on `&mut [i32]`
+   |     method `sort` is available on `Vec<i32>`
    |
 note: method `sort_by_key` modifies its receiver in-place, it is not meant to be used in method chains.
   --> $DIR/chain-method-call-mutation-in-place.rs:3:53

